# 03 - Customer Layer Final

In this notebook the artificially generated data will be joined to the real data

## Import Libraries & Data

In [38]:
import pandas as pd

pd.set_option('display.max_columns', None)

In [39]:
path = '/Users/antoniooliveira/Downloads/NTT project'

art = pd.read_csv(f'{path}/artificial data.csv')
data = pd.read_csv(f'{path}/Gold/Customer_info_dataset.csv')

## Preprocessing

Renaming Gender Variable

In [40]:
art['gender'] = art['gender(s)']
art = art.drop(['gender(s)'], axis = 1)

data['gender'] = data['gender(s)']
data = data.drop(['gender(s)'], axis = 1)

Transforming *has_kids* in *artifical_data* to binary 

In [41]:
art['has_kids'] = art['has_kids'].replace({
    True: 1,
    False: 0,
})

Drop Unnamed Column

In [42]:
data = data.drop(['Unnamed: 0'], axis =1)
data.head(1)

household_key                                           quantity  \
0              1  [1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, ...   

   retail_disc  coupon_disc  coupon_match_disc  \
0    -0.403613    -0.046647          -0.015142   

                                      sales_value_eu  \
0  [4.27, 3.2, 1.17, 3.97, 2.99, 7.69, 2.68, 1.59...   

                                          department first_transaction_date  \
0  ["Groceries", "Groceries", "Groceries", "Delic...             2021-02-20   

  marital_status_code homeowner_desc marital_status age_group  \
0                   A      Homeowner        married    senior   

   adult_category_size  has_kids  avg_age  avg_income  n_kids  n_household  \
0                  2.0       0.0     65.0     42000.0     0.0          2.0   

   Groceries_spend  Delicacies_spend  Meat_spend  Bakery_spend  DRUG GM_spend  \
0          3285.02            226.99      373.56        102.01         568.23   

   SALAD BAR_spend  MISC SALES TRAN_spend  RESTAURANT_spend  FLORAL_spend  \
0            44.09                   21.4              4.47          8.55   

   Seafood_spend  COSMETICS_spend  KIOSK-GAS_spend  CHEF SHOPPE_spend  \
0            0.0              0.0              0.0                0.0   

   GARDEN CENTER_spend  MISC. TRANS._spend  SPIRITS_spend  AUTOMOTIVE_spend  \
0                  0.0                 0.0            0.0               0.0   

   TRAVEL & LEISUR_spend  CNTRL/STORE SUP_spend  COUP/STR & MFG_spend  \
0                    0.0                    0.0                   0.0   

   GM MERCH EXP_spend  POSTAL CENTER_spend  DELI/SNACK BAR_spend  \
0                 0.0                  0.0                   0.0   

   Photo/Video_spend  CHARITABLE CONT_spend  RX_spend  VIDEO RENTAL_spend  \
0                0.0                    0.0       0.0                 0.0   

   PROD-WHS SALES_spend  PHARMACY SUPPLY_spend  TOYS_spend  HBC_spend  \
0                   0.0                    0.0         0.0        0.0   

   ELECT &PLUMBING_spend  HOUSEWARES_spend  gender  
0                    0.0               0.0     2.0

Checking for Missing Values

In [43]:
data.isna().sum()

household_key                0
quantity                     0
retail_disc                  0
coupon_disc                  0
coupon_match_disc            0
sales_value_eu               0
department                   0
first_transaction_date       0
marital_status_code       1699
homeowner_desc            1699
marital_status            1699
age_group                 1699
adult_category_size       1699
has_kids                  1699
avg_age                   1699
avg_income                1699
n_kids                    1699
n_household               1699
Groceries_spend              0
Delicacies_spend             0
Meat_spend                   0
Bakery_spend                 0
DRUG GM_spend                0
SALAD BAR_spend              0
MISC SALES TRAN_spend        0
RESTAURANT_spend             0
FLORAL_spend                 0
Seafood_spend                0
COSMETICS_spend              0
KIOSK-GAS_spend              0
CHEF SHOPPE_spend            0
GARDEN CENTER_spend          0
MISC. TR

As expected, there are 1699 missing values. These will be inputed by joining data from the *artificial_data* dataset

In [44]:
art_iterator = iter(art.itertuples(index=False))

# Fill missing values in `data`
for row_index, row in data.iterrows():
    # Check if the row has missing values
    if row.isna().any():
        # Get the corresponding fill values from `art`
        art_row = next(art_iterator)
        
        # Fill missing values
        for column in data.columns:
            if pd.isna(row[column]):
                data.at[row_index, column] = getattr(art_row, column)

data.head(3)

household_key                                           quantity  \
0              1  [1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, ...   
1              2  [1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 2, 1, 1, 1, ...   
2              3  [1, 1, 1, 1, 3, 2, 2, 1, 1, 1, 1, 2, 1, 1, 1, ...   

   retail_disc  coupon_disc  coupon_match_disc  \
0    -0.403613    -0.046647          -0.015142   
1    -0.469174    -0.012605           0.000000   
2    -0.732278    -0.066367          -0.021475   

                                      sales_value_eu  \
0  [4.27, 3.2, 1.17, 3.97, 2.99, 7.69, 2.68, 1.59...   
1  [0.7, 4.78, 1.07, 0.85, 2.68, 2.68, 2.68, 5.35...   
2  [3.3, 1.75, 1.38, 1.06, 5.35, 2.23, 4.28, 1.81...   

                                          department first_transaction_date  \
0  ["Groceries", "Groceries", "Groceries", "Delic...             2021-02-20   
1  ["Groceries", "Groceries", "Groceries", "Groce...             2021-04-13   
2  ["Groceries", "Groceries", "Groceries", "Groce...             2021-04-23   

  marital_status_code  homeowner_desc marital_status    age_group  \
0                   A       Homeowner        married       senior   
1                   B  Probable Owner         single  young-adult   
2                   A       Homeowner        married  middle-aged   

   adult_category_size  has_kids  avg_age  avg_income  n_kids  n_household  \
0                  2.0       0.0     65.0     42000.0     0.0          2.0   
1                  1.0       0.0     21.5     25000.0     0.0          1.0   
2                  2.0       1.0     49.5    137000.0     2.0          4.0   

   Groceries_spend  Delicacies_spend  Meat_spend  Bakery_spend  DRUG GM_spend  \
0          3285.02            226.99      373.56        102.01         568.23   
1          1361.50             50.01      238.86         26.55         353.11   
2          1982.59             39.02      608.90          3.52         170.16   

   SALAD BAR_spend  MISC SALES TRAN_spend  RESTAURANT_spend  FLORAL_spend  \
0            44.09                  21.40              4.47          8.55   
1             0.00                   4.56              0.00         23.53   
2             1.46                   0.00              0.00          0.00   

   Seafood_spend  COSMETICS_spend  KIOSK-GAS_spend  CHEF SHOPPE_spend  \
0           0.00             0.00             0.00                0.0   
1           9.62            23.62             0.00                0.0   
2          10.68             1.34            21.41                0.0   

   GARDEN CENTER_spend  MISC. TRANS._spend  SPIRITS_spend  AUTOMOTIVE_spend  \
0                  0.0                 0.0            0.0               0.0   
1                  0.0                 0.0            0.0               0.0   
2                  0.0                 0.0            0.0               0.0   

   TRAVEL & LEISUR_spend  CNTRL/STORE SUP_spend  COUP/STR & MFG_spend  \
0                    0.0                    0.0                   0.0   
1                    0.0                    0.0                   0.0   
2                    0.0                    0.0                   0.0   

   GM MERCH EXP_spend  POSTAL CENTER_spend  DELI/SNACK BAR_spend  \
0                 0.0                  0.0                   0.0   
1                 0.0                  0.0                   0.0   
2                 0.0                  0.0                   0.0   

   Photo/Video_spend  CHARITABLE CONT_spend  RX_spend  VIDEO RENTAL_spend  \
0                0.0                    0.0       0.0                 0.0   
1                0.0                    0.0       0.0                 0.0   
2                0.0                    0.0       0.0                 0.0   

   PROD-WHS SALES_spend  PHARMACY SUPPLY_spend  TOYS_spend  HBC_spend  \
0                   0.0                    0.0         0.0        0.0   
1                   0.0                    0.0         0.0        0.0   
2                   0.0                    0.0 

Verifying the success of the operation

In [45]:
data.isna().sum()

household_key             0
quantity                  0
retail_disc               0
coupon_disc               0
coupon_match_disc         0
sales_value_eu            0
department                0
first_transaction_date    0
marital_status_code       0
homeowner_desc            0
marital_status            0
age_group                 0
adult_category_size       0
has_kids                  0
avg_age                   0
avg_income                0
n_kids                    0
n_household               0
Groceries_spend           0
Delicacies_spend          0
Meat_spend                0
Bakery_spend              0
DRUG GM_spend             0
SALAD BAR_spend           0
MISC SALES TRAN_spend     0
RESTAURANT_spend          0
FLORAL_spend              0
Seafood_spend             0
COSMETICS_spend           0
KIOSK-GAS_spend           0
CHEF SHOPPE_spend         0
GARDEN CENTER_spend       0
MISC. TRANS._spend        0
SPIRITS_spend             0
AUTOMOTIVE_spend          0
TRAVEL & LEISUR_spen

Rectify the age_group misscomputed by Gretal.AI in the *artificial data* dataset

In [46]:
for column in data.columns:
    if data[column].nunique() < 20:
        print(data[column].unique())
        print('-----')

['A' 'B' 'U']
-----
['Homeowner' 'Probable Owner' 'Renter' 'Unknown' 'Probable Renter' 'Owner'
 'Probable Renters']
-----
['married' 'single' 'unknown']
-----
['senior' 'young-adult' 'middle-aged' 'adult' 'young adult' 'mid-aged'
 'junior' 'young' 'mid' 'old' 'middle' 'young_adult' 'junior_adult'
 'middle_aged' 'senior_adult' 'middle_age']
-----
[2. 1.]
-----
[0. 1.]
-----
[65.  21.5 49.5 29.5 39.5 59.5]
-----
[ 42000.  25000. 137000.  19500. 162000.  62000.  29500.  87000.  15000.
 112000. 250000. 187000. 224500. 195000.  49500.]
-----
[0. 2. 1. 3.]
-----
[2. 1. 4. 3. 5.]
-----
[ 0.    0.11 21.4   1.07  4.28 10.65  3.21]
-----
[0.   1.61 1.59 1.07 0.35 2.02 9.58 3.17 0.71 2.47 1.69]
-----
[ 0.    4.27  1.75  4.23  3.2   1.2   1.67  3.54 13.58  0.9 ]
-----
[ 0.    8.55  2.14 14.97  3.41  2.66 13.69  5.33  1.07  2.13  2.4   2.  ]
-----
[0.   1.07 7.21]
-----
[ 0.    6.41 22.35 11.76  8.55 10.68  1.27 12.84  6.3  10.69]
-----
[ 0.    6.4  11.22  5.35]
-----
[0.   5.35 2.7 ]
-----
[0.   2

In [47]:
def map_age_group(avg_age):
    if avg_age == 21.5:
        return 'young adult'
    elif avg_age == 29.5:
        return 'adult'
    elif avg_age in [39.5, 49.5]:
        return 'middle-aged'
    else:
        return 'senior'

# Apply the function to create the new column
data['age_group'] = data['avg_age'].apply(map_age_group)

Verify the success of the operation

In [48]:
for column in data.columns:
    if data[column].nunique() < 20:
        print(data[column].unique())
        print('-----')

['A' 'B' 'U']
-----
['Homeowner' 'Probable Owner' 'Renter' 'Unknown' 'Probable Renter' 'Owner'
 'Probable Renters']
-----
['married' 'single' 'unknown']
-----
['senior' 'young adult' 'middle-aged' 'adult']
-----
[2. 1.]
-----
[0. 1.]
-----
[65.  21.5 49.5 29.5 39.5 59.5]
-----
[ 42000.  25000. 137000.  19500. 162000.  62000.  29500.  87000.  15000.
 112000. 250000. 187000. 224500. 195000.  49500.]
-----
[0. 2. 1. 3.]
-----
[2. 1. 4. 3. 5.]
-----
[ 0.    0.11 21.4   1.07  4.28 10.65  3.21]
-----
[0.   1.61 1.59 1.07 0.35 2.02 9.58 3.17 0.71 2.47 1.69]
-----
[ 0.    4.27  1.75  4.23  3.2   1.2   1.67  3.54 13.58  0.9 ]
-----
[ 0.    8.55  2.14 14.97  3.41  2.66 13.69  5.33  1.07  2.13  2.4   2.  ]
-----
[0.   1.07 7.21]
-----
[ 0.    6.41 22.35 11.76  8.55 10.68  1.27 12.84  6.3  10.69]
-----
[ 0.    6.4  11.22  5.35]
-----
[0.   5.35 2.7 ]
-----
[0.   2.13 5.33 6.39 4.26]
-----
[0.   1.59 6.81 1.06]
-----
[0.   2.88 2.16]
-----
[0.   1.07]
-----
[0.  3.2]
-----
[2. 0. 1.]
-----


## Export

In [49]:
data.to_csv(f'{path}/Gold/final_customer_info.csv', index=False)